### PR TITLE
encryption: encrypt user external accounts

### PR DESCRIPTION
**Pull request by asdine**
This PR modifies the user external account store to encrypt user data if a key is configured.
It adds a new column to the `user_external_accounts` to determine if the user data is already encrypted or not,
and automatically decrypt it.

### CODE REVIEW

File: internal/encryption/keyring/ring.go [hunk 1]
```diff
@@ -98,3 +104,39 @@ func NewKey(ctx context.Context, k *schema.EncryptionKey) (encryption.Key, error
 		return nil, fmt.Errorf("couldn't configure key: %v", *k)
 	}
 }
+
+// MaybeEncrypt encrypts data with the given key returns the id of the key. If the key is nil, it returns the data unchanged.
+func MaybeEncrypt(ctx context.Context, key encryption.Key, data string) (maybeEncryptedData, keyID string, err error) {
```
Comment (arussellsaw) [inline internal/encryption/keyring/ring.go:109]: i think i'd rather keep methods like this out of the encryption packages, especially ones that return plaintext secrets not wrapped in encryption.Secret, as the best practices are to keep secrets wrapped in encryption.Secret until they're absolutely needed.

**Review by arussellsaw — COMMENTED** [review]
My main feedback on this is that i think we're doing this encryption/decryption at the wrong level. **This is my fault** for A) not writing documentation, B) having the first implementation be one that goes against the ideal patterns. We decided to embed encryption in the store for ExternalServices because it was really difficult to do anywhere else, but in an ideal world we want to keep data encrypted for as long as possible, and only ever decrypt it when it's used (eg passed as plaintext to authenticate a request).  So i'd move the decryption to be inside the methods that read data directly from the account data blob.

This serves two purposes: 
* Firstly this reduces the probability of accidental disclosure, eg leaking in logs. this is something i've had happen in the past, and it really sucks, so building tooling that makes it harder to accidentally leak is worth the investment.
* Secondly this means anything that reads user_external_accounts now needs access to the encryption key, so a security model of least privilege is harder to achieve, eg any host that reads this row, even if it doesn't need the secure data must now have access to the encryption keys for external accounts.

**However, pragmatically this code is still good**, and we should merge it once we're happy it's reviewed. It is still infinitely better than what we have today, and we shouldn't let perfect get in the way of good. I'll make sure to write documentation so that future people working on these problems can head in the right direction.

**Comment by asdine** [comment]
@arussellsaw I totally agree, we need to improve this in the near future. At least now this data is encrypted at rest, which is already a big step

File: internal/encryption/keyring/ring.go (hunk 1 above)
Comment (asdine) [reply internal/encryption/keyring/ring.go:109]: Good idea, I will keep these helpers in the database package for now

**Comment by unknwon** [comment]
@asdine Do you want another pair of eyes as-is or you would like to address @arussellsaw's suggestion first?

**Comment by asdine** [comment]
@unknwon Yes please feel free to review before I submit the changes, for now the only pending changes are about moving `MaybeEncrypt` to the database package.

File: internal/database/external_accounts.go [hunk 2]
```diff
@@ -69,6 +77,64 @@ func (s *UserExternalAccountsStore) ensureStore() {
...
+}
+
+// maybeEncryptAccountData encrypts and returns user auth data if an encryption.Key is configured
+func (s *UserExternalAccountsStore) maybeEncryptAccountData(ctx context.Context, data extsvc.AccountData) (*encryptedAccountData, string, error) {
+	var enc encryptedAccountData
+	var keyIdent string
+	var err error
+
+	if data.AuthData != nil {
+		enc.AuthData, keyIdent, err = keyring.MaybeEncrypt(ctx, s.getEncryptionKey(), string(*data.AuthData))
```
Comment (unknwon) [inline internal/database/external_accounts.go:98]: I find it much easier to understand if we just call it `keyID`

**Review by unknwon — APPROVED** [review]

**Review by arussellsaw — APPROVED** [review]

**Pull request merged**